src/menus/EditMenus.cpp

```diff
@@ -73,65 +79,173 @@ bool DoPasteText(AudacityProject &project)
+      while(dstTrack != dstTrackRange.end() && srcTrack != srcTrackRange.end())
+      {
+         if(!(*dstTrack)->GetSelected())
+         {
+            ++dstTrack;
+            continue;
+         }
+         
+         auto srcChannelCount = TrackList::Channels(*srcTrack).size();
+         auto dstChannelCount = TrackList::Channels(*dstTrack).size();
+                  
+         while(srcChannelCount > 0 && dstChannelCount > 0)
+         {
+            f(**srcTrack);
+            
+            lastCopiedTrack = srcTrack;
+            ++srcTrack;
+            ++dstTrack;
+            --srcChannelCount;
+            --dstChannelCount;
+         }
+         
+         while(dstChannelCount > 0)
+         {
+            f(**lastCopiedTrack);
+            ++dstTrack;
+            --dstChannelCount;
+         }
+      }
+      while(dstTrack != dstTrackRange.end())
+      {
+         if((*dstTrack)->GetSelected())
+            f(**lastCopiedTrack);
+         ++dstTrack;
+      }
+   }
+}
+
+wxULongLong EstimateCopyBytesCount(const TrackList& src, const TrackList& dst)
+{
+   wxULongLong result{};
+   ForEachCopiedWaveTrack(src, dst, [&](const WaveTrack& waveTrack) {
+      sampleCount samplesCount = 0;
+      for(auto& clip : waveTrack.GetClips())
+         samplesCount += clip->GetSequenceSamplesCount();
+      result += samplesCount.as_long_long() * SAMPLE_SIZE(waveTrack.GetSampleFormat());
+   });
+   return result;
+}
+
+BlockArray::size_type EstimateCopiedBlocks(const TrackList& src, const TrackList& dst)
+{
+   BlockArray::size_type result{};
+   ForEachCopiedWaveTrack(src, dst, [&](const WaveTrack& waveTrack) {
+      for(auto& clip : waveTrack.GetClips())
+         result += clip->GetSequenceBlockArray()->size();
+   });
+   return result;
+}
+
+std::shared_ptr<TrackList> DuplicateDiscardTrimmed(const TrackList& src) {
+   auto result = TrackList::Create(nullptr);
+   for(auto track : src)
+   {
+      auto trackCopy = track->Copy(track->GetStartTime(), track->GetEndTime(), false);
+      trackCopy->Init(*track);
+      trackCopy->SetOffset(track->GetOffset());
+      
+      if(auto waveTrack = dynamic_cast<WaveTrack*>(trackCopy.get()))
+      {
+         for(auto clip : waveTrack->GetClips())
+         {
+            if(clip->GetTrimLeft() != 0)
+            {
+               auto t0 = clip->GetPlayStartTime();
+               clip->SetTrimLeft(0);
+               clip->ClearLeft(t0);
+            }
+            if(clip->GetTrimRight() != 0)
+            {
+               auto t1 = clip->GetPlayEndTime();
+               clip->SetTrimRight(0);
+               clip->ClearRight(t1);
+            }
+         }
+      }
+      result->Add(trackCopy);
+   }
+   return result;
+}
+
+// Create and paste into NEW tracks.
 // (This was formerly the second part of overly-long OnPaste.)
-bool DoPasteNothingSelected(AudacityProject &project)
+void DoPasteNothingSelected(AudacityProject &project, const TrackList& src, double t0, double t1)
```

Comment:
The third commit has no errors, but I don't yet see what code duplication is removed as the commit comment says.  Or rather, this is allowing something to be reused?

Reply:
Clipboard access and double checking on same condition (`!tracks.Selected()`) - `DoPasteNotingSelected` assumes that nothing is selected... but I agree that commit message isn't very clear, I'll change it to `Audio paste code simplification`